### DISCUSSION TITLE
configure issue

### DISCUSSION THREAD

**Discussion by efrainhr** [peridigm/peridigm#275] [Q&A] [answered]
Hi all

New to peridynamics and recently managed to install peridigm on our HPC. I was able to run the examples in peridigm/examples, but failed most (165) of the unit tests.

I think the main issue is related to how I set my paths, because my bin only contains the Peridigm executable

> $ ls $HOME/peridigm/build/bin
> Peridigm

Hence, the unit tests is linking to missing executables, e.g.,

> $ ls -l $HOME/peridigm/build/scripts/epu
> lrwxrwxrwx 1 efrainhr default 11 Oct 18 01:57 $HOME/peridigm/build/scripts/epu -> /../bin/epu

However, these executables are in the Trilinos build. I've tried updating my `$PATH` to include Trilinos/bin and other things, but nothing seems to work.

Here's my cmake script for building peridigm

> #!/bin/bash
>
> module load cseinit-intel
module load cse/lapack
module load cse/cmake
module load cse/hdf5
module load cse/boost
module load cse/netcdf
>
> export CC=$MPI_HOME/bin/mpicc
export CXX=$MPI_HOME/bin/mpicxx
export F77=$MPI_HOME/bin/mpif77
LINKER_FLAGS="-fopenmp "
>
> export LIBRARY_PATH=$NETCDF_LIB_DIR:$LIBRARY_PATH
export LD_LIBRARY_PATH=$NETCDF_LIB_DIR:$LD_LIBRARY_PATH
export CPATH=$NETCDF_INCLUDE_DIR:$CPATH
>
> rm -rf CMake*
>
> cmake -D CMAKE_INSTALL_PREFIX:PATH=$HOME/peridigm/build \
-D CMAKE_BUILD_TYPE:STRING=Release \
-D Trilinos_DIR:PATH=$HOME/Trilinos/build/lib/cmake/Trilinos \
-D CMAKE_C_COMPILER:STRING=$CC \
-D CMAKE_CXX_COMPILER:STRING=$CXX \
-D CMAKE_CXX_FLAGS:STRING="-O2 -Wall -pedantic -Wno-long-long -Wno-deprecated -fPIC -Wl,-Bdynamic" \
-D CMAKE_C_FLAGS:STRING="-O2 -Wall -fPIC -Wl,-Bdynamic" \
$HOME/peridigm/src
>
make -j 4
make install

Using the latest version of Trilinos and peridigm.

Any advice? Thanks!

**Comment by oldninja** [comment]
Can you share some of the error messages on the tests? That usually gives us a good direction where the problem may be at...

**Reply by efrainhr** [reply]
Sorry, but it's not letting me attach the `LastTest.log`.

Here's an example. 

> Traceback (most recent call last):
  File "./np1/SimpleImplicitDiffusion.py", line 48, in <module>
    p = Popen(command, stdout=logfile, stderr=logfile)
  File "$HOME/anaconda/3.8/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "$HOME/anaconda/3.8/lib/python3.8/subprocess.py", line 1706, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: '../../../../scripts/exodiff'

The scripts in that dir point to the `peridigm/bin`, which only includes the `Peridigm` executable. Digging a bit into it, the issue might be how I built Trilinos? 

> $ grep -E "set\(Trilinos_LIB|set\(Trilinos_INC" $HOME/Trilinos/build/lib/cmake/Trilinos/TrilinosConfig.cmake 
set(Trilinos_INCLUDE_DIRS "$HOME/Trilinos/build/include")
set(Trilinos_LIBRARY_DIRS "")

Although, right above those lines it says `# Deprecated (see #299)!`. Really don't know, just guessing at this point. 

Note: just to be clear, I'm able to successfully run the examples in `peridigm/examples`.

**Comment by andrisfreimanis** [comment, marked as answer]
I might have had a similar problem a while ago when I updated to a newer version of Trilinos. The problem was that the CMake makes links from binaries (epu etc.) in Trilinos directory to the /scripts directory. But CMake couldn't find the Trilinos directory. Then the links weren't made and testing suite couldn't use exodiff, so only unit tests passed.

Check the top level CMakeLists.txt line 311-317. CMakeuses `${Trilinos_BINARY_PATH}` to create links. It is defined on line 62 as `SET(Trilinos_BINARY_PATH ${Trilinos_LIBRARY_DIRS}/../bin)`. My issue was that `${Trilinos_LIBRARY_DIRS}` wasn't defined in Trilinos CMake files. I modified line 62 to `SET(Trilinos_BINARY_PATH ${Trilinos_INCLUDE_DIRS}/../bin)` and it worked, because `${Trilinos_INCLUDE_DIRS}` was defined.

**Reply by efrainhr** [reply]
Haven't tested this, but it seems like it would fix the bad linking issues.Thanks!

**Reply by sarthaks20** [reply]
There is one more issue I am facing, trilinos13.0.1/lib/cmake/Trilinos/TrilinosConfig.cmake does not contain anything about Trilinos TPL/external_packages like Netcdf, HDF5, yaml-cpp, etc. which results in error that TPLs are not found while building Peridigm. 
Please help.
Thank you.
